Fixes for builder.sh

gentoo-root opened the pull request:
See the commit messages. The goal is to fix cache paths in builder.sh.

joestringer reviewed (approved):
This change makes sense to me.

Main question I would have is which scenarios we need to test. From your posts I take it that you have validated this for yourself locally. In [this comment](https://github.com/cilium/cilium/pull/37688#issuecomment-2736068323) I saw a previous regression on the command `make GO='contrib/scripts/builder.sh go' generate-apis`. That _seems_ like it matched the failure in CI when renovate runs this script (related PR: https://github.com/cilium/cilium/pull/38423). Not sure if there's anything else we should try out before merging this, or we could just merge this and if there is any fallout we just address those issues as they arise.

joestringer commented on contrib/scripts/builder.sh:
```diff
@@ -6,54 +6,47 @@ cd "$(dirname "$0")/../.."
 if [ -z "${RUN_AS_ROOT:-}" ]; then
-    USER_OPTION="--user $(id -u):$(id -g)"
-    USER_PATH="$HOME"
+	USER_OPTION=(--user "$(id -u):$(id -g)")
```
> An open question is how to handle host user IDs different from 0 and 1000

I think this $USER_OPTION variable already handles that case?

joestringer replied:
Oh I see, ${USER_PATH} (now typically hardcoded to the uid 1000 'ubuntu' user from the builder image) is now the mount path but if you pass a different ID then I guess it picks a different user. Not sure how I thought docker worked but in the past I just naively found that passing `--user` seemed to fix whatever permissions problems I had. I didn't really think about how it implements that functionality and whether it's relying on some assumption about the UID mappings inside of the container image.

I guess if we want to work around this the natural thing would be to create the container, create the target user inside the container, then execute the build(?)

gentoo-root committed:
contrib: Safer arguments passing in builder.sh

Some arguments of docker run are passed unquoted, and are generated from
the uncontrolled output of external commands, which may contain spaces.
To be on the safe side, build those arguments as arrays, each element
quoted separately.

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

gentoo-root committed:
contrib: Fix wrong mount paths in builder.sh

Volume mounts for local runs of builder.sh used wrong paths, both inside
and outside the container, which effectively invalidated the idea of
sharing the caches with the host.

1. The mountpoints inside the container should use the paths configured
for tools inside the container. Instead, the host caches were mounted
at the same paths inside the container, where the container tools
couldn't find them.

2. The ccache directory ~/.ccache is a non-standard location. Ccache
uses $CCACHE_DIR, or the one set in its config file, or falls back to
$XDG_CACHE_DIR/ccache, with some more shenanigans on macOS. Two things
to make it work: a) query the host ccache (if installed) for the actual
cache dir; b) use the correct path inside the container.'

3. The home directory inside the container is either /root or
/home/ubuntu. Set $USER_PATH correctly, instead of using the host $HOME
and mounting the caches to a wrong location in the container. It also
obsoletes the macOS workaround that didn't have a chance to work,
because /tmp/.ccache didn't exist on the host and wasn't used by the
container.

Fixes: 8f02631f230c ("contrib: Reuse local go cache in builder")
Fixes: d582b6a08e56 ("contrib: Use tmp directory for macOS builder")
Signed-off-by: Alice Mikityanska <alice@isovalent.com>

gentoo-root committed:
bpf/tests: Separate build and run stages of run_bpf_tests

Currently, `make run_bpf_tests` spins up a privileged Docker container
to build and run the BPF tests. The container is needed to provide the
right version of LLVM toolchain, the same as used by Cilium deployments.

However, building the tests doesn't require root privileges, and running
them doesn't have to be in a container. Split the build and run stages,
so that building happens as a regular user (and the work directory is
not filled with files owned by root), and running happens outside of
Docker, using sudo to gain root privileges.

The run target of bpf/tests/Makefile was removed, and its contents were
moved to the root Makefile, because bpf/tests/Makefile includes *.d
files generated inside the container, and the absolute paths in these
files differ from the host paths. `go test` runs outside the container
now, and uses sudo to gain root privileges.

A few deprecated options like COVER and INSTRLOG, which are not used,
are removed, but instead the user is provided with an option to pass any
options to `go test` via BPF_TEST_FLAGS.

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

gentoo-root committed:
.github/workflows: Install tparse and go-junit-report in lint-bpf-checks

Since the build step of `make run_bpf_tests` now runs outside of the
Docker container, it needs tparse and go-junit-report installed on the
host, so that junit files could be generated, otherwise Makefile.defs
defines GOTEST_FORMATTER = cat.

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

gentoo-root committed:
Makefile: Use shell quoting for JUNIT_PATH

JUNIT_PATH can be provided by the user and contain double quotes. A new
function shellquote is added to apply shell-style quoting to Makefile
variables when forming a command line. It's supposed to work like @Q in
BASH. Apply the shellquote function to JUNIT_PATH to pass it safely via
GOTEST_FORMATTER.

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

gentoo-root committed:
contrib: Fix UID != 1000 in builder.sh

If the host user is not 1000, it will cause permission errors in the
container, where /home/ubuntu is owned by 1000:1000. Change UID:GID of
the ubuntu user and ownership of the home directory, according to the
host user ID to allow seamless interaction with host-mounted paths.

Host directories for volumes mounted in Docker must be created as the
needed user if they don't exist. Otherwise, Docker would create them as
root and make them unwritable inside the container as a normal user.
This issue existed before, but was masked in GitHub runners (user 1001):
as the target user 1001 didn't exist inside the container, HOME was
detected as /, and GOCACHE used /.cache/go-build, with /.cache
precreated with mode 0777 while building the container. It "worked", but
it didn't actually reuse the host GOCACHE mounted at
/home/ubuntu/.cache/go-build (before this patch) or even at the host
GOCACHE path (before this series).

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

gentoo-root committed:
contrib: Remove RUN_AS_ROOT from builder.sh

The only usage is in the build-go-caches workflow, and it chowns to the
GitHub runner user anyway right after building the cache as root. Save
the extra step and remove unneeded functionality.

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

tklauser commented on .github/workflows/lint-bpf-checks.yaml:
```diff
@@ -127,6 +127,18 @@ jobs:
+        uses: actions/setup-go@d35c59abb061a4a6fb18e82ac0862c26744d6ab5 # v5.5.0
+        with:
+          # renovate: datasource=golang-version depName=go
+          go-version: 1.24.6
```
nit: update to 1.26.1 in case you need to address other review comments as well. Or just let renovate take care of it on the next run 🤷

bimmlerd commented on Makefile:
```diff
@@ -597,19 +597,24 @@ mcs-api-conformance: ## Run MCS-API conformance tests.
+	contrib/scripts/builder.sh \
+		env MAKEFLAGS="$(filter-out --jobserver-auth=%,$(MAKEFLAGS))" \
+		make $(SUBMAKEOPTS) -C bpf/tests/ all
+	$(SUDO) $(GO) test ./bpf/tests/bpftest \
```
Can we use `go test -exec 'sudo -E'` here instead? I think otherwise you might in theory be running a different `go`, depending on `PATH` of the user vs of root. Or is there a reason you _want_ to run go with `sudo`?

bimmlerd commented on contrib/scripts/builder.sh:
```diff
@@ -63,10 +58,8 @@ CONTAINER=$(docker create \
-	docker exec "$CONTAINER" chown -Rhc --from=:1000 :ubuntu /home/ubuntu
-	docker exec "$CONTAINER" usermod -u "$USERID" ubuntu
-fi
+docker exec "$CONTAINER" groupmod -g "$GROUPID" ubuntu
```
Does this work for macOS? I'll let the person with the macbook chime in, but I thought you'd get 20 as your `$GROUPID`; which would then collide with whatever is 20 on ubuntu, no?

tklauser reviewed (approved):
Looks good to me once David's comments are addressed. One very minor nit inline.

gentoo-root committed:
contrib: Reassign used UIDs and GIDs in builder.sh

If the host UID or GID maps to an existing user or group in the
container, reassign its IDs inside the container to avoid clashes and
unexpected side effects.

A practical example: on macOS, users have GID = 20, which maps to group
dialout in the Ubuntu container.

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

gentoo-root committed:
contrib: chown ~/.cache in builder.sh

Although the code that mounts the ~/.cache/go-build volume into the
container takes measures to ensure the right ownership, ~/.cache doesn't
exist in the container, and Docker creates it as root. Chown it
manually.

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

gentoo-root committed:
contrib: Workaround for BASH 3.2 in builder.sh

macOS uses BASH 3.2 that has a bug and treats an empty array as "unbound
variable". This was fixed in BASH 4.4, but macOS still uses an older
version. Add a workaround that temporarily sets +u for the affected
commands. The rest of variables used in these commands (PWD and
CILIUM_BUILDER_IMAGE) are always set.

Signed-off-by: Alice Mikityanska <alice@isovalent.com>

bimmlerd reviewed (approved):
Thanks!
